pipelinewise/fastsync/commons/tap_postgres.py

```diff
@@ -348,18 +349,33 @@ def fetch_current_log_pos(self):
             raise Exception('Logical replication not supported before PostgreSQL 9.4')
 
         try:
-            # Create replication slot.
-            self.create_replication_slot()
+            # Create replication slot and obtain the oldest LSN
+            # that it can deliver.
+            oldest_lsn = self.create_replication_slot()
         except psycopg2.errors.DuplicateObject:
-            # Replication slot exists already so continue.
-            pass
+            # Replication slot exists already so just get the
+            # oldest LSN which the slot can deliver.
+            oldest_lsn = self.get_confirmed_flush_lsn()
 
         # Close replication slot dedicated connection
         self.primary_host_conn.close()
 
-        lsn = self.get_current_lsn()
+        current_lsn = self.get_current_lsn()
 
-        return {'lsn': lsn, 'version': 1}
+        if self.connection_config.get('replica_host'):
+            # Ensure that the newest LSN availiable on the replica is newer than
+            # the oldest LSN which is deliverable by the replication slot.
+            time_waited = 0  # seconds
+            wait_period = 1  # seconds
+            max_time_waited = 60  # seconds
+            while oldest_lsn > current_lsn:
+                time.sleep(wait_period)
+                time_waited += wait_period
+                if time_waited > max_time_waited:
+                    raise RuntimeError('Replica database is lagging too far behind Primary.')
+                current_lsn = self.get_current_lsn()
```

Comment:
Just to check -- can the oldest LSN change while this is happening? Or is that pinned once the connection is started?

Reply:
`oldest_lsn` is really `oldest_lsn_which_is_required_for_consistent_replication` which is either provided by Postgres when the replication slot is created or required as the most recent `confirmed_flush_lsn` for the slot. `confirmed_flush_lsn` cannot change while this is happening as PipelineWise is the only application which should be talking to this slot and it is busy doing this.